integrations/rust-project/src/sysroot.rs

```diff
@@ -112,16 +120,82 @@ pub(crate) fn resolve_rustup_sysroot() -> Result<Sysroot, anyhow::Error> {
     let mut output = utf8_output(cmd.output(), &cmd)?;
     truncate_line_ending(&mut output);
     let sysroot = PathBuf::from(output);
-    let sysroot_src = sysroot
-        .join("lib")
-        .join("rustlib")
-        .join("src")
-        .join("rust")
-        .join("library");
 
-    let sysroot = Sysroot {
-        sysroot,
-        sysroot_src: Some(sysroot_src),
+    let sysroot = if let Some(sysroot_src) = sysroot_src_override {
+        validate(&sysroot_src)
+            .context("Invalid --sysroot-src, did not contain the standard library source code")?;
+        Sysroot {
+            sysroot,
+            sysroot_src,
+        }
+    } else {
+        let sysroot_src = Sysroot::sysroot_src_for_sysroot(&sysroot);
+        validate(&sysroot_src)
+            .context("Rustup toolchain did not have rust-src component installed")?;
+        Sysroot {
+            sysroot_src,
+            sysroot,
+        }
     };
+
     Ok(sysroot)
 }
+
+pub(crate) fn resolve_provided_sysroot(
+    sysroot: &Path,
+    sysroot_src_override: Option<&Path>,
+    project_root: &Path,
+    relative_paths: bool,
+) -> Result<Sysroot, anyhow::Error> {
+    let mut sysroot = expand_tilde(sysroot)?.canonicalize().context(format!(
+        "--sysroot path could not be canonicalized: {}",
+        sysroot.display()
+    ))?;
+    let mut sysroot_src = if let Some(path) = sysroot_src_override {
+        let path = expand_tilde(path)?.canonicalize().context(format!(
+            "--sysroot-src path could not be canonicalized: {}",
+            path.display()
+        ))?;
+        validate(&path)
+            .context("Invalid --sysroot-src, did not contain the standard library source code")?;
+        path
+    } else {
+        let path = Sysroot::sysroot_src_for_sysroot(&sysroot);
+        validate(&path)
+            .context("Provided --sysroot did not contain the standard library source code")?;
+        path
+    };
+    if relative_paths {
+        sysroot = relative_to(&sysroot, project_root);
+        sysroot_src = relative_to(&sysroot_src, project_root);
+    }
+    Ok(Sysroot {
+        sysroot,
+        sysroot_src,
+    })
+}
+
+fn validate(sysroot_src: &Path) -> Result<(), anyhow::Error> {
```

Comment:
naive question: how would you feel about reducing this to a `rustup component list` and scanning for a `rust-src`? rustup doesn't really have machine-readable output, but that might simplify this a bit...

Reply:
Doesn't work if you use Nix without rustup. Which I do. This way works for all methods including, for example, a custom sysroot built by buck.